File: verilog/dv/cocotb/user_proj_tests/uart0/uart0.c
// SPDX-FileCopyrightText: 2023 Efabless Corporation

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//      http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


#include <firmware_apis.h>
#include "user.h"
#include "EF_UART.c"
#include "EF_UART.h"


void main(){
    // Enable managment gpio as output to use as indicator for finishing configuration  
    ManagmentGpio_outputEnable();
    ManagmentGpio_write(0);
    enableHkSpi(0); // disable housekeeping spi
    // configure all gpios as  user out then chenge gpios from 32 to 37 before loading this configurations
    User_enableIF(); // this necessary when reading or writing between wishbone and user project if interface isn't enabled no ack would be recieve and the command will be stuck
    GPIOs_configureAll(GPIO_MODE_USER_STD_BIDIRECTIONAL);
    GPIOs_loadConfigs(); // load the configuration 
    // configure la [63:32] as output from cpu
    EF_DRIVER_STATUS status;
    status = UART_Init(UART0_USER, 2083333, 50000000, 8, true, ODD, 0x3F, 0, 10);
    ManagmentGpio_write(1); // configuration finished 

    if (status != EF_DRIVER_OK) {
        ManagmentGpio_write(0);
        return -1;
    }
    // Receive a message
    char buffer[100];
    status = EF_UART_readCharArr(UART0_USER, buffer, sizeof(buffer));
    if (status == EF_DRIVER_OK) {
        // Print received message
        // Transmit a message received
        status = EF_UART_writeCharArr(UART0_USER, buffer);
        if (status != EF_DRIVER_OK) {
            // Handle transmission error
            ManagmentGpio_write(0);
            return -1;
        }
    } else {
        // Handle reception error
        ManagmentGpio_write(0);
        return -1;
    }
    ManagmentGpio_write(0);
    return 0;

}

File: verilog/dv/cocotb/user_proj_tests/uart0/uart0.py
# SPDX-FileCopyrightText: 2023 Efabless Corporation

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#      http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# SPDX-License-Identifier: Apache-2.0


from caravel_cocotb.caravel_interfaces import test_configure
from caravel_cocotb.caravel_interfaces import report_test
import cocotb
from user_proj_tests.uart import ML_UART_BASE
import string
import random

@cocotb.test()
@report_test
async def uart0(dut):
    caravelEnv = await test_configure(dut, timeout_cycles=159844)

    cocotb.log.info(f"[TEST] Start counter_la test")  
    # wait for start of sending
    await caravelEnv.release_csb()
    uart = UART_USER(caravelEnv, prescaler=2)
    await caravelEnv.wait_mgmt_gpio(1)
    cocotb.log.info(f"[TEST] finish configuration") 
    # generate rondom msg
    chars = string.ascii_letters + string.digits
    msg = ''.join(random.choice(chars) for _ in range(10))
    cocotb.log.info(f"[TEST] random msg = {msg}")
    # send random msg
    await uart.uart_send_line(msg)
    # recieve random msg
    msg_rec = await uart.get_line()
    if msg != msg_rec:
        cocotb.log.error(f"[TEST] test failed, recieved msg != sent msg, recieved: {msg_rec}, sent: {msg}")
    await caravelEnv.wait_mgmt_gpio(0)


class UART_USER(ML_UART_BASE):
    def __init__(self, caravelEnv, uart_pins={"tx": 13, "rx": 12}, prescaler=0):
        super().__init__(caravelEnv, uart_pins, prescaler)
        pass
